[JEP-234] Customizable header

imonteroperez opened the pull request:
<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://www.jenkins.io/download/lts/#backporting-process for more.
-->
See [JEP-234](https://github.com/jenkinsci/jep/tree/master/jep/234)

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* New extension point `Header` as an interface that provides capabilities to render a specific header and a default implementation of that, named `JenkinsHeader` that is enabled by default.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://www.jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@MarkEWaite @jglick @fcojfernandez @raul-arabaolaza @EstherAF @Wadeck @timja @batmat 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

imonteroperez committed:
Jenkins header revamp proposal

imonteroperez committed:
Use invokeStatic instead of Functions

imonteroperez committed:
Remove headerDocumentation

imonteroperez committed:
Remove headerDocumentation

timja commented:
Can you write a proper changelog entry rather than a developer guide please, our changelog isn't setup for something that complex.

This should be probably be documented on jenkins.io too?

timja commented on core/src/main/java/jenkins/views/JenkinsHeader.java:
```diff
@@ -0,0 +1,38 @@
+    @CheckForNull
+    public static Header get() {
+        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).collect(Collectors.toList());
+        if (headers.size() > 0) {
```
```suggestion
        if (!headers.isEmpty()) {
```

imonteroperez replied:
Updated in https://github.com/jenkinsci/jenkins/pull/5909/commits/88eac160e4ec72ee1a60680d1a8f36da8c79242e

timja commented on core/src/main/java/jenkins/views/JenkinsHeader.java:
```diff
@@ -0,0 +1,38 @@
+        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).collect(Collectors.toList());
+        if (headers.size() > 0) {
+            if (headers.size() > 1) {
+                LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
```
how does the user review? is this logging useful to an administrator?

timja replied:
also a bit confused isn't the point of this that you want to ship 2 headers, the jenkins one and the cloudbees one, and you want yours to have a higher priority

imonteroperez replied:
> isn't the point of this that you want to ship 2 headers, the jenkins one and the cloudbees one, and you want yours to have a higher priority

JEP proposes to have just one header loaded in our Jenkins instance. IOW proposal is about to make it feasible to have enough customization techniques from Jenkins to create headers, but we want to limit the number of active implementations of that header to one on runtime. Does it make sense to have more than one header at the same time active in the same Jenkins instance? I do not think so. Do we have the capabilities to limit the headers active? Yes, if header is an extension point we can use an [ExtensionFilter#allows](https://javadoc.jenkins.io/jenkins/ExtensionFilter.html#allows-java.lang.Class-hudson.ExtensionComponent-) method for that purpose.

> how does the user review? is this logging useful to an administrator?
This is shown in the Jenkins log if more than one headers are available. I think it is useful for admins to know that there exist two or more headers active.

Vlatombe replied:
Why not use a similar aproach to what is currently done for `UpdateCenter` or `PluginManager` (with an override class  [defined](https://github.com/jenkinsci/jenkins/blob/57737eec99d13341b986aeb6db02dda17d2aaf48/core/src/main/java/hudson/PluginManager.java#L185) through a system property?) As you said, it makes no sense to have multiple headers enabled at one point in time, and this customization is likely desired at OEM level (CloudBees shipping its own war, or an enterprise providing its own header implementation through system property).

imonteroperez replied:
I did not consider that approach. BTW I would prefer doing this transparently without the need of setting extra/additional system properties.

imonteroperez replied:
All comments addressed. Finally no need to use ExtensionFilter, also getting only the one with top priority and issues reported on warnings and admin monitor. Regarding the alternative approach, as mentioned, preferred not to set up extra/additional system properties.

timja commented on core/src/main/resources/jenkins/views/JenkinsHeader/headerContent.jelly:
```diff
@@ -0,0 +1,80 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
+	<header id="header" class="page-header">
```
don't mix tabs and spaces

imonteroperez replied:
Fixed in https://github.com/jenkinsci/jenkins/pull/5909/commits/284abf7c8fdcc2da6e16e72e8c9a9ec33895aaab

timja commented on core/src/main/resources/jenkins/views/JenkinsHeader/headerContent.jelly:
```diff
@@ -0,0 +1,80 @@
+            </a>
+        </div>
+
+        <div class="searchbox hidden-xs">
```
biggest concern with this PR is about making this header an API, what happens when we want to change it and it breaks plugins that have customised it.

Could you show a reference implementation that shows a realistic scenario and the changes that were needed to be done.

i.e. would it be better to create specific APIs along the lines of:

* Logo
* Area to the right of the logo
* ...

imonteroperez replied:
A reference implementation is included in the JEP description. See [JEP-234](https://github.com/imonteroperez/jep/blob/jep-401-header/jep/234/README.adoc)

imonteroperez replied:
See: https://github.com/imonteroperez/custom-header-plugin/tree/main/src/main

jglick replied:
Agreed that this is the main concern. https://github.com/jenkinsci/jep/pull/380#discussion_r749337871 Is the current design suitable only for headers which totally replace the stock GUI with a completely different style? Seems to be impossible to incrementally modify particular pieces. Possibly OK if so, but needs to be emphasized in the JEP and **Reasoning** explaining this choice.

imonteroperez replied:
JEP updated on reasoning section https://github.com/jenkinsci/jep/pull/380#commits-pushed-2499ba3

oleg-nenashev commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,13 @@
+
+import hudson.ExtensionPoint;
+
+public interface Header extends ExtensionPoint {
```
Needs Javadoc for the new extension point

imonteroperez replied:
Included in https://github.com/jenkinsci/jenkins/pull/5909/commits/88eac160e4ec72ee1a60680d1a8f36da8c79242e

oleg-nenashev commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,13 @@
+public interface Header extends ExtensionPoint {
+    
+    /**
+     * Checks if header is enabled. By default it is if installed, but the logic is deferred in the plugins.
```
Wrong comment. There is no default behavior

imonteroperez replied:
You are right. Updated in https://github.com/jenkinsci/jenkins/pull/5909/commits/88eac160e4ec72ee1a60680d1a8f36da8c79242e

oleg-nenashev reviewed (dismissed):
I do not think this is a correct approach for such a change. As @timja mentioned, this makes maintenance more complicated. And it is not an extension point in classic means, it is just an override logic. It might be helpful for some parties if they want to replace the header alltogether bit it complicates further maintenance. As long as there is no reference implementation justifying it, I am not ready to support such design.

imonteroperez committed:
Simplification

imonteroperez committed:
Some feedback from PR

imonteroperez committed:
Merge branch 'header-revamp' of github.com:imonteroperez/jenkins into header-revamp

imonteroperez committed:
Fix ident

imonteroperez commented:
> I do not think this is a correct approach for such a change. As @timja mentioned, this makes maintenance more complicated. And it is not an extension point in classic means, it is just an override logic. It might be helpful for some parties if they want to replace the header alltogether bit it complicates further maintenance. As long as there is no reference implementation justifying it, I am not ready to support such design.

@oleg-nenashev in [JEP-234](https://github.com/imonteroperez/jep/blob/jep-401-header/jep/234/README.adoc) there is a reference implementation. Not sure if this is enough or not, but as mentioned in JEP, current alternatives make no possible (or a hacky/hell/nightmare-frankenstein-thing) to have a customized header (not only in UI terms)

timja commented:
> > I do not think this is a correct approach for such a change. As @timja mentioned, this makes maintenance more complicated. And it is not an extension point in classic means, it is just an override logic. It might be helpful for some parties if they want to replace the header alltogether bit it complicates further maintenance. As long as there is no reference implementation justifying it, I am not ready to support such design.
> 
> @oleg-nenashev in [JEP-234](https://github.com/imonteroperez/jep/blob/jep-401-header/jep/234/README.adoc?rgh-link-date=2021-11-11T15%3A46%3A46Z) there is a reference implementation. Not sure if this is enough or not, but as mentioned in JEP, current alternatives make no possible (or a hacky/hell/nightmare-frankenstein-thing) to have a customized header (not only in UI terms)

This is exactly what I was concerned with:

https://github.com/imonteroperez/custom-header-plugin/blob/main/src/main/resources/org/jenkinsci/plugins/custom/header/CustomHeader/headerContent.jelly

a narrower API would be preferred I think

imonteroperez commented:
> > > I do not think this is a correct approach for such a change. As @timja mentioned, this makes maintenance more complicated. And it is not an extension point in classic means, it is just an override logic. It might be helpful for some parties if they want to replace the header alltogether bit it complicates further maintenance. As long as there is no reference implementation justifying it, I am not ready to support such design.
> > 
> > 
> > @oleg-nenashev in [JEP-234](https://github.com/imonteroperez/jep/blob/jep-401-header/jep/234/README.adoc?rgh-link-date=2021-11-11T15%3A46%3A46Z) there is a reference implementation. Not sure if this is enough or not, but as mentioned in JEP, current alternatives make no possible (or a hacky/hell/nightmare-frankenstein-thing) to have a customized header (not only in UI terms)
> 
> This is exactly what I was concerned with:
> 
> https://github.com/imonteroperez/custom-header-plugin/blob/main/src/main/resources/org/jenkinsci/plugins/custom/header/CustomHeader/headerContent.jelly
> 
> a narrower API would be preferred I think

No strong opinion on to setup API endpoints for each part of the header content (logo, searchbar, etc.) to make it more modular, but not part of the scope of the JEP proposed here. JEP aims to have capabilities to include/update parts/all the header beyond styles and static content. I guess I get your point and I see that as a follow up of this in order to apply KISS :-)

timja commented:
Again as Oleg and I have said this isn't extending this is replacing.

This does not explain backwards compatibility for Jenkins core in terms of the header:
https://github.com/imonteroperez/jep/blob/jep-401-header/jep/234/README.adoc#backwards-compatibility

If we refactor the header what happens to your custom header, could be completely broken.
If you want to continue down this path it needs to be clearly documented as a non stable API as this sort of API could never be stable imo.

we could change CSS and adapt the header but your custom one would not be adapted at all.

imonteroperez commented:
> Again as Oleg and I have said this isn't extending this is replacing.

I'm fine with that statement. I have not mentioned at all it is about extending. Actually, as you mentioned, the JEP proposal aims to make it feasible to customize by a replacement of the `pageHeader.jelly` that should look like:

```xml
<?jelly escape-by-default='true'?>
<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
	<j:invokeStatic var="header" className="jenkins.views.JenkinsHeader" method="get"/>
	<st:include it="${header}" page="headerContent.jelly"/>
</j:jelly>
```
And each plugin that provides a `Header` implementation can replace the `headerContent`. The main benefit here is that this open the door to have capabilities in Jenkins to customize the header beyond UI and static content (like CSS), so we would be able to do things like:

```xml
<j:invokeStatic var="label" className="org.jenkinsci.plugins.custom.header.CustomHeader" method="getHeaderLabel"/>
<span class="hidden-xs hidden-sm">${label} - ${userName}</span>
```

delegating to Java to programmatically retrieve things to be rendered in the header or to add new business logic there.

> This does not explain backwards compatibility for Jenkins core in terms of the header: https://github.com/imonteroperez/jep/blob/jep-401-header/jep/234/README.adoc#backwards-compatibility
> 
> If we refactor the header what happens to your custom header, could be completely broken. If you want to continue down this path it needs to be clearly documented as a non stable API as this sort of API could never be stable imo.
> 
> we could change CSS and adapt the header but your custom one would not be adapted at all.

Well, I guess this is the same issue that could happen with other approaches like simple-theme-plugin. For example: if we decide tomorrow to modify the id of the search-box on the header: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/layout/pageHeader.jelly#L53, this will probably affect a lot of themes using that customization tooling (i.e: https://github.com/djonsson/jenkins-atlassian-theme/blob/master/src/style.css#L95)

IOW (and setting expectations) as mentioned before I'm ok with having a narrowed API that could help here to fix the mentioned issues, but that was not the proposal that is on the JEP. The proposal of the JEP aims to have capabilities to include/update parts/all the header beyond styles and static content.

imonteroperez commented:
BTW JEP is open for discussion here: https://github.com/jenkinsci/jep/pull/380 and if it is leading to confusion in terms of scope (like extending vs. replacing) I'm happy to update it to reflect the desired scope and improvement proposed there.

imonteroperez committed:
Merge remote-tracking branch 'upstream/master' into header-revamp

imonteroperez committed:
Fix conflicts

timja commented:
> Well, I guess this is the same issue that could happen with other approaches like simple-theme-plugin. For example: if we decide tomorrow to modify the id of the search-box on the header

Themes are specifically documented as no API compatibility:
https://www.jenkins.io/doc/book/managing/ui-themes/#themes-support-policy

imonteroperez commented:
> > Well, I guess this is the same issue that could happen with other approaches like simple-theme-plugin. For example: if we decide tomorrow to modify the id of the search-box on the header
> 
> Themes are specifically documented as no API compatibility: https://www.jenkins.io/doc/book/managing/ui-themes/#themes-support-policy

Got it, then I will update the JEP section about backward compatibility to make it explicit. Thanks for the heads up and great feedback @timja!

imonteroperez committed:
Delete pageHeader.jelly.orig

timja commented:
happier with the concept given the additional backwards compat notes in the jep, why is this draft?

imonteroperez commented:
> happier with the concept given the additional backwards compat notes in the jep, why is this draft?

:smile: Great! No strong reason. I will open the PR as ready to review.

amuniz commented on core/src/main/java/jenkins/views/JenkinsHeader.java:
```diff
@@ -0,0 +1,44 @@
+    
+    @Restricted(NoExternalUse.class)
+    @CheckForNull
+    public static Header get() {
```
Don't this method belongs to the `Header` interface (which should be an abstract class then)? 
`JenkinsHeader` should be _just_ another extension, which happens to be always there.

imonteroperez replied:
Sounds good. Updated in https://github.com/jenkinsci/jenkins/pull/5909/commits/ae940851ff1ec57f2d214a35d3626e8280fac23e. I let in the `JenkinsHeader` before given it was the default implementation of the `Header` provided by the core, but yes, I prefer to include it there and to refactor as an abstract class instead of an interface.

amuniz commented on core/src/main/resources/lib/layout/pageHeader.jelly:
```diff
@@ -25,82 +25,6 @@
-            </j:if>
-        </div>
-    </header>
+	<j:invokeStatic var="header" className="jenkins.views.JenkinsHeader" method="get"/>
```
Related to the previous comment, I think this should be:
```
<j:invokeStatic var="header" className="jenkins.views.Header" method="get"/>
```
So it's not tied to an implementation of the extension point.

imonteroperez replied:
Updated in https://github.com/jenkinsci/jenkins/pull/5909/commits/ae940851ff1ec57f2d214a35d3626e8280fac23e

imonteroperez committed:
Move method get to Header

imonteroperez committed:
Merge branch 'header-revamp' of github.com:imonteroperez/jenkins into header-revamp

imonteroperez committed:
Remove extra space

imonteroperez committed:
Setup visibility properly

jtnord commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,43 @@
+        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).collect(Collectors.toList());
+        if (!headers.isEmpty()) {
+            if (headers.size() > 1) {
+                LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
```
but the `JenkinsHeader` is always present and hard coded to be enabled so if you try and have another Header you will always have 2.

if your implementation / extension needs to also provide an ExtensionFilter to work (which it seems like it) then why fallback to the Jenkins header, this will completely and utterly spam the logs for every page load IIUC.
Or alternatively do not use an ExtensionFilter to remove the base JenkinsHeader and just use the Ordinal (and then you can just create and install one, and it will be some much simpler here - just return `ExtensionList.lookup(Header.class).get(0)`

imonteroperez replied:
:thinking: how is this going to work when there exist two or more headers with the same `ordinal` value? (Consider two headers with ordinal as `Integer.MAX_VALUE`). I would prefer to give plugin developers the capability to filter the extension to be loaded to ensure its header is going to be the one active and rendered. Does it make sense?

jtnord replied:
> ). I would prefer to give plugin developers the capability to filter the extension to be loaded to ensure its header is going to be the one active and rendered. Does it make sense?

not to me. :)

Because right now both plugin developers are *forced* to write an `ExtentionFilter`, they would likely just write one that removed all `Header` implementations that where not their own... thus you would even fallback to the `JenkinsHeader` that was itself filtered so not an extension (and you create a new one for each page view) and not even warn the user why as the headers extensionList would be empty.

If you did not force them to do that - you would likely not get into that situation to begin with., they would still be free to do that if they desire - but in all reality this would to me be better handled by warning the user in the UI to remove some plugin.

imonteroperez replied:
Got it, IUCC then the main concern is not about making it feasible to use the `ExtensionFilter` at all, it is about current implementation is forcing developers to use it. Good point. Then I will just rely on the `ordinal` value as proposed. Updated in: https://github.com/jenkinsci/jenkins/pull/5909/commits/6fe15520b3b7fa327c0c7ba89cf3f67da5f46ce9

jtnord reviewed (commented):
Just a thought:

As there is some concern about breaking API compatability why not introduce a field in the Header.

when an incompatable change is made in the `Header` (like the search form API), the version can be increased.  If the version number of the extension `Header` is not the version number required then do not use the customization, or log the warning at startup, disaplay an administrative monitor etc..

This means that it will be obvious something has changed,  can be checked at test time (e.g. using the PCT on the extension plugin, or an ATH would obviously fail for said plugin).

imonteroperez committed:
Do not force to use ExtensionFilter

jtnord commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,36 @@
+    @Restricted(NoExternalUse.class)
+    @CheckForNull
+    public static Header get() {
+        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).collect(Collectors.toList());
+        if (headers.isEmpty()) {
+            return new JenkinsHeader();
+        }
+        return headers.get(0);
+    }
```
`headers` should never be empty - you always have the fallback `JenkinsHeader` `Extension` with lowest ordinal that is always enabled?

If you do not have that Jenkins is broken (by some plugin code or other so I think it is perfectly fine to go 💥 )?  

in any case there is no need to `collect` here - just `findFirst()` should suffice?

Something like (with the correct imports) ?

```suggestion
        Optional<Header> header = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).findFirst());
        // return header.orElseThrow(() -> new IllegalStateException("something is very wrong here"));
        return header.orElseGet(() -> new JenkinsHeader());
        
    }
```

imonteroperez replied:
Applied in https://github.com/jenkinsci/jenkins/pull/5909/commits/1dd8d32e52489255cc7e91e25b226be138995f82

jglick commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,36 @@
+    @Restricted(NoExternalUse.class)
+    @CheckForNull
+    public static Header get() {
+        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(header -> header.isHeaderEnabled()).collect(Collectors.toList());
```
```suggestion
        List<Header> headers = ExtensionList.lookup(Header.class).stream().filter(Header::isHeaderEnabled).collect(Collectors.toList());
```

plus @jtnord’s suggestions

imonteroperez replied:
Applied in https://github.com/jenkinsci/jenkins/pull/5909/commits/1dd8d32e52489255cc7e91e25b226be138995f82

jglick commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,36 @@
+     * Checks if header is enabled.
+     * @return if header is enabled
+     */
+    public abstract boolean isHeaderEnabled();
```
```suggestion
    public abstract boolean isEnabled();
```

jglick commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,36 @@
+ * @author Ildefonso Montero
+ * @see JenkinsHeader
+ */
+public abstract class Header implements ExtensionPoint {
```
`interface` cleaner I think.

imonteroperez replied:
Changed from `interface` to `abstract class` to apply: https://github.com/jenkinsci/jenkins/pull/5909#discussion_r749171416

jglick commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,36 @@
+ * 
+ * @author Ildefonso Montero
+ * @see JenkinsHeader
+ */
```
```suggestion
 * @since TODO
 */
```

imonteroperez replied:
Applied in https://github.com/jenkinsci/jenkins/pull/5909/commits/1dd8d32e52489255cc7e91e25b226be138995f82

jglick commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Extension point that provides capabilities to render a specific header
+ * 
+ * @author Ildefonso Montero
```
```suggestion
```

`git blame` if anyone ever cares.

imonteroperez replied:
Applied in https://github.com/jenkinsci/jenkins/pull/5909/commits/1dd8d32e52489255cc7e91e25b226be138995f82

imonteroperez committed:
Update get method

jtnord commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,32 @@
+    public abstract boolean isEnabled();
+
+    @Restricted(NoExternalUse.class)
+    @CheckForNull
```
can never be null with the current code path.
```suggestion
```

imonteroperez committed:
Update core/src/main/java/jenkins/views/Header.java

Co-authored-by: James Nord <jtnord@users.noreply.github.com>

jglick commented on core/src/main/java/jenkins/views/JenkinsHeader.java:
```diff
@@ -5,14 +5,13 @@
- * @author Ildefonso Montero
  * @see Header
  */
 @Extension(ordinal = Integer.MIN_VALUE)
```
Or just
```suggestion
```
since you are already https://github.com/jenkinsci/jenkins/blob/1dd8d32e52489255cc7e91e25b226be138995f82/core/src/main/java/jenkins/views/Header.java#L30

jtnord replied:
with the code as it was this will for the OSS "header is not extended" use case this just return a cached object, whereas now it will create a new `JenkinsHeader` unnecessarily on every call.

imonteroperez committed:
Update core/src/main/java/jenkins/views/JenkinsHeader.java

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

imonteroperez committed:
Merge remote-tracking branch 'upstream/master' into header-revamp

imonteroperez committed:
Fix some idents

amuniz commented on core/src/main/java/jenkins/views/JenkinsHeader.java:
```diff
@@ -0,0 +1,15 @@
+ * 
+ * @see Header
+ */
+public class JenkinsHeader extends Header {
```
Missing `@Extension`?

amuniz replied:
Oh, I see.. closing.

imonteroperez committed:
Deal with compatibility (partial/full) headers

imonteroperez commented:
> Just a thought:
> 
> As there is some concern about breaking API compatability why not introduce a field in the Header.
> 
> when an incompatable change is made in the `Header` (like the search form API), the version can be increased. If the version number of the extension `Header` is not the version number required then do not use the customization, or log the warning at startup, disaplay an administrative monitor etc..
> 
> This means that it will be obvious something has changed, can be checked at test time (e.g. using the PCT on the extension plugin, or an ATH would obviously fail for said plugin).

@jtnord @jglick @timja 

Updated code to deal with backward compatibility based on @jtnord proposal. As mentioned on JEP there are two different cases to manage:
- Consider placing all your required CSS and Javascript code inside your custom plugins if you are going to do a complete refactor of the header. 
- Consider being up-to-date with the latest sources/updates on the `headerContent` in case you were doing minimal changes through your custom header plugin.

For these two scenarios, I created two specific headers, `FullHeader` which is going to be totally independent and will not rely on references to core resources such as images, CSS elements, etc. and `PartialHeader` which is used to perform minimal changes and relies on core resources references.

For the latest one, we need to check if it is compatible or not using a specific version number that should be increased on introducing breaking/incompatible changes.

Updated in: https://github.com/jenkinsci/jenkins/pull/5909/commits/10e26c6406a27f54e3192980ab6e2313fe2af990

Tested and working as expected forcing custom plugin to be non-compatible
```
2021-11-16 12:19:27.361+0000 [id=50]	INFO	jenkins.InitReactorRunner$1#onAttained: System config adapted
2021-11-16 12:19:27.362+0000 [id=50]	INFO	jenkins.InitReactorRunner$1#onAttained: Loaded all jobs
2021-11-16 12:19:27.363+0000 [id=42]	WARNING	jenkins.views.PartialHeader#incompatibleHeaders: org.jenkinsci.plugins.custom.header.CustomHeader:0.1 not compatible with 1.0
```

I will update the JEP content to consider this on Backward compatibility section.

imonteroperez committed:
Fix build

jtnord commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,44 @@
+    
+    @Initializer(after = InitMilestone.JOB_LOADED, before = InitMilestone.JOB_CONFIG_ADAPTED)
+    public static void incompatibleHeaders() {
+        for (PartialHeader header: ExtensionList.lookup(PartialHeader.class).stream().filter(h -> !h.isCompatible()).collect(Collectors.toList())) {
+            LOGGER.warn(String.format("%s:%s not compatible with %s", header.getClass().getName(), header.getVersion(), compatibilityHeaderVersion));
+        }
```
also use an `AdministrativeError` so the message does not get lost (no one checks the logs!) ?

```suggestion
        for (PartialHeader header: ExtensionList.lookup(PartialHeader.class).stream().filter(h -> !h.isCompatible()).collect(Collectors.toList())) {
            LOGGER.warn(String.format("%s:%s not compatible with %s", header.getClass().getName(), header.getVersion(), compatibilityHeaderVersion));
            new AdministrativeError(header.getClass().getName(), "Incompatible Header", String.format("The plugin %s is attempting to replace the Header but is not compatible with this version of Jenkins.  The plugin should be updated or can be removed." Jenkins.get().getPluginManager().whichPlugin(header.getClass()), null);
        }
```

imonteroperez replied:
Updated in 82b68da41ddd91d6fd13c48bd0858ce1b9ae5d38

imonteroperez committed:
Use administrative monitor and use int instead of VersionNumber

imonteroperez commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * @return the header version
+     */
+    public abstract VersionNumber getVersion();
```
Updated in 82b68da41ddd91d6fd13c48bd0858ce1b9ae5d38

imonteroperez commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,44 @@
+    
+    @Override
+    public boolean isCompatible() {
+        return compatibilityHeaderVersion.isOlderThanOrEqualTo(getVersion());
```
Updated in 82b68da41ddd91d6fd13c48bd0858ce1b9ae5d38

imonteroperez commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,44 @@
+    private static Logger LOGGER = Logger.getLogger(PartialHeader.class.getName());
+    
+    /** When an incompatible change is made in the header (like the search form API), compatibility header version should be increased */
+    private static final VersionNumber compatibilityHeaderVersion = new VersionNumber("1.0");
```
Updated in 82b68da41ddd91d6fd13c48bd0858ce1b9ae5d38

imonteroperez commented:
@jtnord tested with latest changes and administrative monitor warning on incompatible headers

![Captura de pantalla de 2021-11-16 15-27-46](https://user-images.githubusercontent.com/595347/142003726-c50e7522-1bd7-4634-afbb-af6c6bcee4ee.png)

```
2021-11-16 14:26:20.181+0000 [id=50]	INFO	jenkins.InitReactorRunner$1#onAttained: System config adapted
2021-11-16 14:26:20.183+0000 [id=52]	INFO	jenkins.InitReactorRunner$1#onAttained: Loaded all jobs
2021-11-16 14:26:20.188+0000 [id=51]	WARNING	jenkins.views.PartialHeader#incompatibleHeaders: org.jenkinsci.plugins.custom.header.CustomHeader:0 not compatible with 1
```

jtnord reviewed (approved):
Thanks for the quick updates @imonteroperez

Looks reasonable to me.

timja commented on core/src/main/resources/lib/layout/pageHeader.jelly:
```diff
@@ -25,82 +25,6 @@
-            </j:if>
-        </div>
-    </header>
+	<j:invokeStatic var="header" className="jenkins.views.Header" method="get"/>
+	<st:include it="${header}" page="headerContent.jelly"/>
```
nit, rest of the file uses spaces

imonteroperez replied:
Tabs removed in https://github.com/jenkinsci/jenkins/pull/5909/commits/68a6b9c2768fb5e328dfcdd1740959a617ffef69. Thanks for the heads up @timja!

imonteroperez committed:
Remove tabs

Vlatombe commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,46 @@
+
+    private static Logger LOGGER = Logger.getLogger(PartialHeader.class.getName());
+    
+    /** When an incompatible change is made in the header (like the search form API), compatibility header version should be increased */
```
```suggestion
    /**
     * The current compatibility version of the Header API.
     * 
     * Increment this number when an incompatible change is made to the header (like the search form API).
     */
```

Vlatombe commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,46 @@
+    private static final int compatibilityHeaderVersion = 1;
+    
+    @Override
+    public boolean isCompatible() {
```
```suggestion
    public final boolean isCompatible() {
```

Vlatombe commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,46 @@
+    public abstract int getSupportedHeaderVersion();
+    
+    @Initializer(after = InitMilestone.JOB_LOADED, before = InitMilestone.JOB_CONFIG_ADAPTED)
+    public static void incompatibleHeaders() {
```
```suggestion
    @SuppressWarnings("unused")
    public static void incompatibleHeaders() {
```

Vlatombe commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,46 @@
+    
+    @Initializer(after = InitMilestone.JOB_LOADED, before = InitMilestone.JOB_CONFIG_ADAPTED)
+    public static void incompatibleHeaders() {
+        for (PartialHeader header: ExtensionList.lookup(PartialHeader.class).stream().filter(h -> !h.isCompatible()).collect(Collectors.toList())) {
+            LOGGER.warn(String.format("%s:%s not compatible with %s", header.getClass().getName(), header.getSupportedHeaderVersion(), compatibilityHeaderVersion));
+            new AdministrativeError(header.getClass().getName(), "Incompatible Header", String.format("The plugin %s is attempting to replace the Header but is not compatible with this version of Jenkins.  The plugin should be updated or can be removed.", Jenkins.get().getPluginManager().whichPlugin(header.getClass())), null);
+        }
```
```suggestion
        ExtensionList.lookup(PartialHeader.class).stream().filter(h -> !h.isCompatible()).forEach(header -> {
            LOGGER.warn(String.format("%s:%s not compatible with %s", header.getClass().getName(), header.getSupportedHeaderVersion(), compatibilityHeaderVersion));
            new AdministrativeError(header.getClass().getName(), "Incompatible Header", String.format("The plugin %s is attempting to replace the Jenkins header but is not compatible with this version of Jenkins. The plugin should be updated or removed.", Jenkins.get().getPluginManager().whichPlugin(header.getClass())), null);
        });
```

Vlatombe commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public abstract class PartialHeader extends Header {
+
+    private static Logger LOGGER = Logger.getLogger(PartialHeader.class.getName());
```
Should be using jul, not log4j.

imonteroperez replied:
Updated in https://github.com/jenkinsci/jenkins/pull/5909/commits/2fa8ea7fb63537b56c48cf390782423004ddf634

Vlatombe commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,45 @@
+import hudson.ExtensionPoint;
+
+/**
+ * Extension point that provides capabilities to render a specific header
```
```suggestion
 * Extension point that provides capabilities to render a specific header.
 *
 * Extend {@link PartialHeader} or {@link FullHeader} depending on the use case.
```

imonteroperez committed:
Apply suggestions from code review

Co-authored-by: Vincent Latombe <vincent@latombe.net>

imonteroperez committed:
Use jul

jtnord commented on core/src/main/java/jenkins/views/JenkinsHeader.java:
```diff
@@ -0,0 +1,15 @@
+ * 
+ * @see Header
+ */
+public class JenkinsHeader extends FullHeader {
```
```suggestion
@Extension(ordinal = Integer.MIN_VALUE)
public class JenkinsHeader extends FullHeader {
```

So we are not creating a new `JenkinsHeader` on every call if the Header is not overridden.

imonteroperez committed:
Update core/src/main/java/jenkins/views/JenkinsHeader.java

Co-authored-by: James Nord <jtnord@users.noreply.github.com>

imonteroperez committed:
Fix import

raul-arabaolaza reviewed (approved):
I believe any useful feedback I can give is already given by previous reviewers so...

fcojfernandez reviewed (approved):
Looks great @imonteroperez

timja commented:
Is the JEP enough documentation for this feature or should it be documented on jenkins.io first?

jglick commented:
> documented on jenkins.io

I would not think so. Documentation should be in Javadoc.

timja commented:
> > documented on jenkins.io
> 
> I would not think so. Documentation should be in Javadoc.

Then it needs more docs, it doesn’t say what jelly files are supported

imonteroperez committed:
Add some javadoc to Header

imonteroperez commented:
> > > documented on jenkins.io
> > 
> > 
> > I would not think so. Documentation should be in Javadoc.
> 
> Then it needs more docs, it doesn’t say what jelly files are supported

@jglick @timja Javadoc improved on the `Header` extension point to explain explicitly how it works and what jelly files are required. See: https://github.com/jenkinsci/jenkins/pull/5909/commits/5ed6ad99e43b2c5a4fa8b9c3678c4a26a0d9bf75

timja commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -0,0 +1,60 @@
+ * The default Jenkins header is provided as an implementation of a {@link FullHeader} 
+ * named {@link JenkinsHeader}.
+ * 
+ * All headers will provide a prioritization technique, via the usual ordinal field of 
+ * the {@link Extension} annotation.
```
something like this maybe.
```suggestion
 * The first header located will be used, set the ordinal field on {@link Extension} to have a higher priority.
```

imonteroperez committed:
Update core/src/main/java/jenkins/views/Header.java

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

imonteroperez committed:
Fix javadoc ref

Wadeck commented on core/src/main/java/jenkins/views/PartialHeader.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    private static final int compatibilityHeaderVersion = 1;
+    
+    @Override
+    public final boolean isCompatible() {
+        return compatibilityHeaderVersion == getSupportedHeaderVersion();
+    }
+    
+    /**
+     * @return the supported header version
+     */
+    public abstract int getSupportedHeaderVersion();
```
Perhaps it's too restrictive to force "equal" for the compatibility. No strong opinion

Wadeck reviewed (approved):
Nice improvement compared to the first draft I have seen!

batmat commented:
> I do not think this is a correct approach for such a change. As @timja mentioned, this makes maintenance more complicated

@oleg-nenashev I think your feedback has been addressed. Also given Tim approved this PR in the meantime, care to re-review?
Thanks!

---

Apart from the comment above, it looks like we're close to ready for merge. I think it would be cleaner however to officially merge the [JEP PR](https://github.com/jenkinsci/jep/pull/380) before merging the core PR here.

@imonteroperez there has been comments by Jesse & Mark on the JEP, can you please make sure they're all addressed and marked as resolved so we can move towards merging there and here? Thanks for the great work!

imonteroperez commented:
> @imonteroperez there has been comments by Jesse & Mark on the JEP, can you please make sure they're all addressed and marked as resolved so we can move towards merging there and here? Thanks for the great work!

@batmat all comments addressed and marked as resolved. Not able to merge the JEP PR (restricted to authorized users)

Vlatombe commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -10,9 +10,22 @@
+ * All headers will provide a prioritization technique, via the usual ordinal field of 
+ * the {@link Extension} annotation.
+ * 
+ * The header content will be injected inside the pageHeader.jelly, based on the header
```
```suggestion
 * The header content will be injected inside the {@code pageHeader.jelly}, based on the header
```

Vlatombe commented on core/src/main/java/jenkins/views/Header.java:
```diff
@@ -10,9 +10,22 @@
+ * 
+ * The header content will be injected inside the pageHeader.jelly, based on the header
+ * retrieved by the {@link Header#get()} method. That header content will be provided
+ * inside a resource called headerContent.jelly. It performs a full replacement
```
```suggestion
 * inside a resource called {@code headerContent.jelly}. It performs a full replacement
```

imonteroperez committed:
Apply suggestions from code review

Co-authored-by: Vincent Latombe <vincent@latombe.net>

batmat commented:
@oleg-nenashev I think your feedback was addressed, so I'm dismissing your past review FYI. Feel free to tell us whether you still have comments or requests on the current implementation.

batmat commented:
Given the 6 approvals, I am marking this PR `ready-for-merge`. 

However, I'd like to condition the merge here to be _after_ https://github.com/jenkinsci/jep/pull/380 has been merged as Draft JEP and assigned an official JEP number. 

In other words, this means I plan to merge here 24 hours from now, _provided_ the JEP PR has been merged in the meantime.

Please tell us if you disagree with this stance. Thanks all.

oleg-nenashev reviewed (dismissed):
Sorry @batmat but I am against merging in the current state. I would like to review it, and AFAICT there was no review requested from me. Sorry, I missed the previous comments but we expect to request reviews in this repository. 

The Tuesday weekly ship has sailed, so there is no problem with keeping the pull request open until the JEP is properly integrated. I will handle that and do the review

imonteroperez commented:
> Sorry @batmat but I am against merging in the current state. I would like to review it, and AFAICT there was no review requested from me. Sorry, I missed the previous comments but we expect to request reviews in this repository.

It was requested 12 days ago, see: https://github.com/jenkinsci/jenkins/pull/5909#event-5604669789

oleg-nenashev commented:
I might be missing something. As far as I can tell it wasn't on my review dashboard. If it was requested through the comment, I can assure you I missed that.

Back to the subject matter, would it be possible to summarize the current review state in the developer mailing list thread? It would be particularly useful to know whether it was presented to the user experience group like it was announced there, and what is the consensus built in the UX SIG.

For the record I do not intend to block the change further. What I wanted is to ensure that we follow the JEP process as much as possible  since it was taken for this pull request. We have one week before the next weekly, so it makes no sense to rush now and to risk creating leftovers

I will follow up on my proposal to add all officers and board members to JEP editors. Apparently even the recent changes are not enough

imonteroperez commented:
> Back to the subject matter, would it be possible to summarize the current review state in the developer mailing list thread? It would be particularly useful to know whether it was presented to the user experience group like it was announced there, and what is the consensus built in the UX SIG.

I will drop all in the dev ML thread tomorrow after the UX SIG meeting

batmat commented:
For the record, the associated JEP for this change here is now _officially_ [JEP-234](https://github.com/jenkinsci/jep/tree/master/jep/234).

imonteroperez commented:
> > Back to the subject matter, would it be possible to summarize the current review state in the developer mailing list thread? It would be particularly useful to know whether it was presented to the user experience group like it was announced there, and what is the consensus built in the UX SIG.
> 
> I will summarize all in the dev ML thread tomorrow after the UX SIG meeting

FTR I shared JEP-234 in the Jenkins UX SIG meeting. Meeting notes are available [here](https://docs.google.com/document/d/1QttPwdimNP_120JukigKsRuBvMr34KZhVfsbgq1HFLM/edit#)

Also shared with them the implementation details from this PR

Feedback obtained from the group was positive and they did not see any issue with this approach. 

Shared this also on the jenkins-dev ML: https://groups.google.com/g/jenkinsci-dev/c/1tDvSioCaF0

cc/ @oleg-nenashev

imonteroperez commented:
Any pending action here before merging this PR? @oleg-nenashev @batmat

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

imonteroperez commented:
> This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.
> 
> Thanks!

@timja Can we merge this? No negative feedback was obtained in 24h.

timja commented:
> > This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.
> > Thanks!
> 
> @timja Can we merge this? No negative feedback was obtained in 24h.

generally we batch merge before weekly 😄  so no need to worry if its more than 24h
